sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java

```diff
@@ -0,0 +1,402 @@
+            Pattern.DOTALL);
+
+    public static OrderByFieldPath fromString(String fieldPath) {
+      if (fieldPath.isEmpty()) {
+        throw new IllegalArgumentException("Could not resolve empty field path");
+      }
+      String originalString = fieldPath;
+      List<String> segments = new ArrayList<>();
+      while (!fieldPath.isEmpty()) {
+        Matcher segmentMatcher = FIELD_PATH_SEGMENT_REGEX.matcher(fieldPath);
+        boolean foundMatch = segmentMatcher.lookingAt();
+        if (!foundMatch) {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        String fieldName;
+        if ((fieldName = segmentMatcher.group(1)) != null) {
+          segments.add(fieldName);
+        } else if ((fieldName = segmentMatcher.group(2)) != null) {
+          String unescaped = unescapeFieldName(fieldName.substring(1, fieldName.length() - 1));
+          segments.add(unescaped);
+        } else {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        fieldPath = fieldPath.substring(fieldName.length());
+        // Due to the regex, any non-empty fieldPath will have a dot before the next nested field.
+        if (fieldPath.startsWith(".")) {
+          fieldPath = fieldPath.substring(1);
+        }
+      }
+      return new OrderByFieldPath(originalString, ImmutableList.copyOf(segments));
+    }
+
+    private final String originalString;
+    private final ImmutableList<String> segments;
+
+    private OrderByFieldPath(String originalString, ImmutableList<String> segments) {
+      this.originalString = originalString;
+      this.segments = segments;
+    }
+
+    public String getOriginalString() {
+      return originalString;
+    }
+
+    public boolean isDocumentName() {
+      return segments.size() == 1 && "__name__".equals(segments.get(0));
+    }
+
+    public ImmutableList<String> getSegments() {
+      return segments;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object other) {
+      if (other instanceof OrderByFieldPath) {
+        return this.segments.equals(((OrderByFieldPath) other).getSegments());
+      }
+      return super.equals(other);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(segments);
+    }
+
+    @Override
+    public int compareTo(OrderByFieldPath other) {
+      // Inspired by com.google.cloud.firestore.FieldPath.
+      int length = Math.min(this.getSegments().size(), other.getSegments().size());
+      for (int i = 0; i < length; i++) {
+        byte[] thisField = this.getSegments().get(i).getBytes(StandardCharsets.UTF_8);
+        byte[] otherField = other.getSegments().get(i).getBytes(StandardCharsets.UTF_8);
+        int bytes_length = Math.min(thisField.length, otherField.length);
+        for (int j = 0; j < bytes_length; j++) {
+          int cmp = Byte.compare(thisField[j], otherField[j]);
```

Comment:
This looks like a signed byte comparator, which would be wrong (try a test with an ascii character vs a UTF8 multi-byte code point).

If you have access to Guava's `UnsignedBytes.lexicographicalComparator` this can be used directly on the byte arrays returned by `getBytes`.

Reply:
Done. Thanks!